src/XMonad/Core.hs

```diff
@@ -447,126 +449,97 @@ runOnWorkspaces job = do
--- used to store temporary files that can easily be recreated.  For
--- example, the XPrompt history file.
---
--- Several directories are considered.  In order of preference:
---
---   1. The directory specified in the @XMONAD_CACHE_DIR@ environment variable.
---   2. The @~\/.xmonad@ directory.
---   3. The @XDG_CACHE_HOME/xmonad@ directory.
---
--- The first directory that exists will be used.  If none of the
--- directories exist then (1) will be used if it is set, otherwise (2)
--- will be used.  Either way, a directory will be created if necessary.
-getXMonadCacheDir :: MonadIO m => m String
-getXMonadCacheDir =
-    findFirstDirWithEnv "XMONAD_CACHE_DIR"
-      [ getAppUserDataDirectory "xmonad"
-      , getXDGDirectory XDGCache "xmonad"
-      ]
-
--- | Return the path to the xmonad data directory.  This directory is
--- used by XMonad to store data files such as the run-time state file
--- and the configuration binary generated by GHC.
---
--- Several directories are considered.  In order of preference:
---
---   1. The directory specified in the @XMONAD_DATA_DIR@ environment variable.
---   2. The @~\/.xmonad@ directory.
---   3. The @XDG_DATA_HOME/xmonad@ directory.
---
--- The first directory that exists will be used.  If none of the
--- directories exist then (1) will be used if it is set, otherwise (2)
--- will be used.  Either way, a directory will be created if necessary.
-getXMonadDataDir :: MonadIO m => m String
-getXMonadDataDir =
-    findFirstDirWithEnv "XMONAD_DATA_DIR"
-      [ getAppUserDataDirectory "xmonad"
-      , getXDGDirectory XDGData "xmonad"
-      ]
-
--- | Helper function that will find the first existing directory and
--- return its path.  If none of the directories can be found, create
--- and return the first from the list.  If the list is empty this
--- function returns the historical @~\/.xmonad@ directory.
-findFirstDirOf :: MonadIO m => [IO FilePath] -> m FilePath
-findFirstDirOf []        = findFirstDirOf [getAppUserDataDirectory "xmonad"]
-findFirstDirOf possibles = do
-    found <- go possibles
-
-    case found of
-      Just path -> return path
-      Nothing   -> do
-        primary <- io (head possibles)
-        io (createDirectoryIfMissing True primary)
-        return primary
+-- | All the directories that xmonad will use.  They will be used for
+-- the following purposes:
+--
+-- * @dataDir@: This directory is used by XMonad to store data files
+-- such as the run-time state file and the configuration binary
+-- generated by GHC.
+--
+-- * @cfgDir@: This directory is where user configuration files are
+-- stored (e.g, the xmonad.hs file).  You may also create a @lib@
+-- subdirectory in the configuration directory and the default recompile
+-- command will add it to the GHC include path.
+--
+-- * @cacheDir@: This directory is used to store temporary files that
+-- can easily be recreated.  For example, the XPrompt history file.
```

Comment:
It's a bit funny that the configuration binary, which can be easily recreated, gets placed in dataDir, whereas the prompt history, which requires superhuman memory abilities or superhuman backup regime to recreate, goes in cacheDir. But xmonad isn't the only program that does it this way so I guess it doesn't matter.